Curl: update win32 support

johnwparent opened the pull request:
Properly specify Curl builder interface for `static` vs `shared` curl with NMake system to ensure all built curls export expected symbols.

Symlinks curl library build artifact to more idiomatic name for `FindCurl.cmake` implementations and other NMake consumers.

johnwparent committed:
curl: update for proper shared vs static build handling on Windows

johnwparent committed:
curl: install artifact link expected name

alecbcs commented on var/spack/repos/builtin/packages/curl/package.py:
```diff
@@ -355,6 +360,13 @@ def flag_handler(self, name, flags):
 
+class BuildEnvironment:
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        if "~shared" in self.spec:
+            env.append_flags("CFLAGS", "-DCURL_STATICLIB")
+            env.append_flags("CXXFLAGS", "-DCURL_STATICLIB")
```
```suggestion
        if self.spec.satisfies("~shared"):
            env.append_flags("CFLAGS", "-DCURL_STATICLIB")
            env.append_flags("CXXFLAGS", "-DCURL_STATICLIB")
```

alecbcs commented on var/spack/repos/builtin/packages/curl/package.py:
```diff
@@ -441,12 +453,12 @@ def with_or_without_secure_transport(self, activated):
     def nmake_args(self):
         args = []
-        mode = "dll" if "libs=dll" in self.spec else "static"
+        mode = "dll" if "+shared" in self.spec else "static"
```
```suggestion
        mode = "dll" if self.spec.satisfies("+shared") else "static"
```

alecbcs commented on var/spack/repos/builtin/packages/curl/package.py:
```diff
@@ -288,6 +288,11 @@ class Curl(NMakePackage, AutotoolsPackage):
             # curl queries pkgconfig for openssl compilation flags
             depends_on("pkgconfig", type="build")
 
+    with when("platform=windows"):
+        # Curl should be built either static or shared on Windows to avoid
+        # collisions between import library and static library
+        variant("shared", default=True, description="Build shared curl")
```
Is there any particular reason to keep this variant as windows specific? It would seem valuable to be able to specify a static build on Linux or macOS as well.

johnwparent replied:
Curl builds both shared and static libs on non Windows platforms, and so this variant is not necessary to discriminate, you just get both.

johnwparent replied:
Oh I see the `libs` variant now, I'll just use that, just doesn't need to be multi valued on Windows.

alecbcs commented on var/spack/repos/builtin/packages/curl/package.py:
```diff
@@ -490,3 +502,12 @@ def install(self, pkg, spec, prefix):
         with working_dir(os.path.join(self.stage.source_path, "builds")):
             install_dir = glob.glob("libcurl-**")[0]
             install_tree(install_dir, self.prefix)
+        if "~shared" in spec:
```
```suggestion
        if spec.satisfies("~shared"):
```

alecbcs reviewed (dismissed):
It's recommended to use `spec.satisfies()` instead of `in spec` these days since the latter can result in unintended consequences see the docs [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#testing-spec-constraints).

Instead of creating a windows specific variant called `shared` I'd recommend seeing if we can generalize this PR to also work on Linux and MacOS.

johnwparent commented:
> It's recommended to use `spec.satisfies()` instead of `in spec` these days since the latter can result in unintended consequences see the docs [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#testing-spec-constraints).

Thanks for the heads up, didn't know that was the idiomatic approach now.

johnwparent committed:
Unify variant accross platforms

johnwparent commented:
@alecbcs I refactored so that all platforms now use the `libs` variant (single valued, default to `shared` on Windows) and updated the spec checks to use `spec.satisfies` with the `libs` variant
